go/parquet/file/column_writer.go

```diff
@@ -0,0 +1,567 @@
+		dp := NewDataPageV1WithStats(memory.NewBufferBytes(data), int32(w.numBuffered), w.encoding, parquet.Encodings.RLE, parquet.Encodings.RLE, uncompressed, pageStats)
+		defer dp.Release()
+		w.WriteDataPage(dp)
+	}
+	return nil
+}
+
+func (w *columnWriter) buildDataPageV2(defLevelsRLESize, repLevelsRLESize, uncompressed int64, values []byte) error {
+	var data []byte
+	if w.pager.HasCompressor() {
+		w.compressedTemp.Reset()
+		data = w.pager.Compress(w.compressedTemp, values)
+		// data = w.compressedTemp.Bytes()
+	} else {
+		data = values
+	}
+
+	// concatenate uncompressed levels and the possibly compressed values
+	combined := make([]byte, int(defLevelsRLESize+repLevelsRLESize+int64(len(data))))
+	copy(combined, w.repLevelSink.Bytes()[:repLevelsRLESize])
+	copy(combined[repLevelsRLESize:], w.defLevelSink.Bytes()[:defLevelsRLESize])
+	copy(combined[repLevelsRLESize+defLevelsRLESize:], data)
+
+	pageStats, err := w.getPageStatistics()
+	if err != nil {
+		return err
+	}
+	pageStats.ApplyStatSizeLimits(int(w.props.MaxStatsSizeFor(w.descr.Path())))
+	pageStats.Signed = schema.SortSIGNED == w.descr.SortOrder()
+	w.resetPageStatistics()
+
+	numValues := int32(w.numBuffered)
+	nullCount := int32(pageStats.NullCount)
+	defLevelsByteLen := int32(defLevelsRLESize)
+	repLevelsByteLen := int32(repLevelsRLESize)
+
+	page := NewDataPageV2WithStats(memory.NewBufferBytes(combined), numValues, nullCount, numValues, w.encoding,
+		defLevelsByteLen, repLevelsByteLen, uncompressed, w.pager.HasCompressor(), pageStats)
+	if w.hasDict && !w.fallback {
+		w.totalCompressedBytes += int64(page.buf.Len()) // + sizeof pageheader
+		w.pages = append(w.pages, page)
+	} else {
+		w.totalCompressedBytes += int64(len(combined))
+		defer page.Release()
+		w.WriteDataPage(page)
+	}
+	return nil
+}
+
+func (w *columnWriter) FlushBufferedDataPages() {
+	if w.numBuffered > 0 {
+		w.AddDataPage()
+	}
+
+	for _, p := range w.pages {
+		defer p.Release()
+		w.WriteDataPage(p)
+	}
+	w.pages = w.pages[:0]
+	w.totalCompressedBytes = 0
+}
+
+func (w *columnWriter) writeLevels(numValues int64, defLevels, repLevels []int16) int64 {
+	toWrite := int64(0)
+	// if the field is required and non-repeated, no definition levels
+	if defLevels != nil && w.descr.MaxDefinitionLevel() > 0 {
+		for _, v := range defLevels {
+			if v == w.descr.MaxDefinitionLevel() {
+				toWrite++
+			}
+		}
+		w.WriteDefinitionLevels(defLevels[:numValues])
+	} else {
+		toWrite = numValues
+	}
+
+	if repLevels != nil && w.descr.MaxRepetitionLevel() > 0 {
+		// a row could include more than one value
+		//count the occasions where we start a new row
+		for _, v := range repLevels {
+			if v == 0 {
+				w.rowsWritten++
+			}
+		}
+
+		w.WriteRepetitionLevels(repLevels[:numValues])
+	} else {
+		// each value is exactly 1 row
+		w.rowsWritten += int(numValues)
+	}
+	return toWrite
+}
+
+func (w *columnWriter) writeLevelsSpaced(numLevels int64, defLevels, repLevels []int16) {
```

Comment:
is there a reason to not re-use the logic above?

Reply:
i guess it avoid  calculated valuestowrite?

Reply:
yea, that's the only reason it's not reusing the logic above it. no need to calculate values to write.